numba/ir_utils.py

```diff
@@ -2020,3 +2020,62 @@ def check_and_legalize_ir(func_ir):
         msg +=  func_ir.diff_str(orig_ir)
         warnings.warn(NumbaWarning(msg, loc=func_ir.loc))
 
+
+def convert_code_obj_to_function(code_obj, caller_ir):
+    """
+    Converts a code object from a `make_function.code` attr in the IR into a
+    python function, caller_ir is the FunctionIR of the caller and is used for
+    the resolution of freevars.
+    """
+    fcode = code_obj.code
+    nfree = len(fcode.co_freevars)
+
+    # try and resolve freevars if they are consts in the caller's IR
+    # these can be baked into the new function
+    freevars = []
+    for x in fcode.co_freevars:
+        freevar_def = caller_ir.get_definition(x)
+        if isinstance(freevar_def, ir.Const):
+            freevars.append(freevar_def.value)
+        else:
+            msg = ("Cannot capture the non-constant value associated with "
+                   "variable '%s' in a function that will escape." % x)
+            raise TypingError(msg, loc=code_obj.loc)
+
+    func_env = "\n".join(["  c_%d = %s" % (i, x) for i, x in enumerate(freevars)])
```

Comment:
The following is a lot like `get_ir_of_code()`. Can the code be de-duplicated?

Reply:
Yes, assuming the fixes for default args etc are ok I'll factor out the stuff that mangles `CodeType`.

Reply:
done